### ISSUE TITLE
Request to have a more granular configuration over `JWT` on `SecurityPolicy`.

### ISSUE THREAD

**Issue by ardikabs** [envoyproxy/gateway#1930]
*Description*:
Security Policy `JWT` is derived from [envoy.extensions.filters.http.jwt_authn.v3.JwtAuthentication](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/jwt_authn/v3/config.proto#envoy-v3-api-msg-extensions-filters-http-jwt-authn-v3-jwtauthentication), in which the extension provide some other functionality to support JWT as authentication. 
Currently, Security Policy `JWT` has some sort of limitations over the actual Envoy Proxy extensions, such as:
1. It only supports the Secured Remote JWKS endpoint (https).
2. It forced the user to follow through with the default behavior for parsing the JWT token via the `Authorization` Header using Bearer format or `access_token` from the query parameter.
3. It couldn't configure `bypass_cors_preflight`.

*Request*:

In my org, we have some requirements that it couldn't follow through the behavior offered from Security Policy `JWT` by EG, thus we added patches to support our requirements such as:
1. Supporting insecure remote JWKS endpoint using http (e.g., port 80), e.g., [patch](https://github.com/ardikabs/gateway/commit/0cc5c50039174e22c9e051e39a0dbb5fdb5e2562#diff-a824c8143caabd2fc1838053a53d28e3bfad1fbd2737427f8c002fd09005f65aR358-R368). The reason why we need to have insecure remote JWKS because we don't have yet the needs to expose the JWKS endpoint, and if so, it will add additional cost for another hop back and forth, like internet->gateway->internet->gateway->internal.
2. Supporting parsing the JWT token from available options such as `from_cookies`, e.g. [patch](https://github.com/ardikabs/gateway/commit/0cc5c50039174e22c9e051e39a0dbb5fdb5e2562#diff-a824c8143caabd2fc1838053a53d28e3bfad1fbd2737427f8c002fd09005f65aR137-R148).
3. We have a plan to set `bypass_cors_preflight` to **true** because we observed that enabling the `JWT` policy will force the CORS to be enabled, while the EG does not yet support configuring the CORS. CMIIW.

If this could be accepted, I would like to work on this enhancement based on the above requirements, at least.

[optional *Relevant Links*:]
1. [Envoy JWT Authn Provider](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/jwt_authn/v3/config.proto#extensions-filters-http-jwt-authn-v3-jwtprovider)

**Comment by arkodg** [comment]
hey @ardikabs, please see inline

> Supporting insecure remote JWKS endpoint using http (e.g., port 80), e.g., [patch](https://github.com/ardikabs/gateway/commit/0cc5c50039174e22c9e051e39a0dbb5fdb5e2562#diff-a824c8143caabd2fc1838053a53d28e3bfad1fbd2737427f8c002fd09005f65aR358-R368). The reason why we need to have insecure remote JWKS because we don't have yet the needs to expose the JWKS endpoint, and if so, it will add additional cost for another hop back and forth, like internet->gateway->internet->gateway->internal.

Imo this doesnt sound like a good idea, it increases the scope of mitm attacks.
Would a `local` provider suffice your needs ?

>Supporting parsing the JWT token from available options such as from_cookies, e.g. [patch](https://github.com/ardikabs/gateway/commit/0cc5c50039174e22c9e051e39a0dbb5fdb5e2562#diff-a824c8143caabd2fc1838053a53d28e3bfad1fbd2737427f8c002fd09005f65aR137-R148).

sounds like a valid use case, any suggestions on user facing API ?

We have a plan to set bypass_cors_preflight to true because we observed that enabling the JWT policy will force the CORS to be enabled, while the EG does not yet support configuring the CORS. CMIIW.

imo this should not be needed since we have the `CORS` field in the `SecurityPolicy`

**Comment by ardikabs** [comment]
> Imo this doesnt sound like a good idea, it increases the scope of mitm attacks.
Would a local provider suffice your needs ?

I don't think a `local` provider is the right solution for my case because It seems using a `local` provider, requires providing a static JWKS file inside the Envoy Proxy pod. About my use case, I still need to access the actual endpoint to fetch the JWKS file, but it is just that the endpoint wasn't exposed to the internet, like an in-cluster access only.

IMO, about the MITM attacks, I don't think that would be a problem because it remains configurable, though using an insecure port must be intended for either a trusted network or an internal network. Like Istio, it also has support for this scenario, when remote JWKS using HTTP scheme or port 80, see [here](https://github.com/istio/istio/blob/937cd21f61c40016595233e6a7c7b25e56c81196/pkg/config/security/security.go#L65).

> sounds like a valid use case, any suggestions on user facing API ?

What I can think right now should look like below:

```
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: jwt-example
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: foo
  jwt:
    providers:
      - name: example
        remoteJWKS:
          uri: https://raw.githubusercontent.com/envoyproxy/gateway/main/examples/kubernetes/jwt/jwks.json
        fromCookies:
          - session_access_token
        fromParams:
          - access_token
        fromHeaders:
          - name: Authorization
            valuePrefix: "Bearer "
```

> imo this should not be needed since we have the CORS field in the SecurityPolicy

Noted, then EG will let `bypass_cors_preflight` be disabled by default.

**Comment by arkodg** [comment]
hey @ardikabs can we split up 2. into a separate issue ?

**Comment by ardikabs** [comment]
@arkodg, We will split between the support for additional parser for JWT token field (eg., `fromCookies`), and the remote jwks support with  plain http (80), correct?

**Comment by arkodg** [comment]
yah @ardikabs to ensure each feature is not blocked by the other
